### DISCUSSION TITLE
Extend d8/v8 to read stdin to TypedArray, or just keep using os.system()?

### DISCUSSION THREAD

**Discussion by guest271314** [quickjs-ng/quickjs#463] [Q&A] [answered]
Right now I'm using QuickJS to read standard input stream to `d8` with `os.system()`, to get around `readline()` not being capable of processing `Uint32Array` from Chrome.

How would I go about translating working QuickJS code to extending `d8` to read standard input into `TypedArray`s?

In `d8` 

```
function encodeMessage(str) {
  return new Uint8Array([...str].map((s) => s.codePointAt()));
}
// ...
    const stdin = os.system("./readFile.js", [
      `/proc/${pid.replace(/\D+/g, "")}/fd/0`,
    ]);
    if (stdin != undefined && stdin != null && stdin.length) {
      const data = encodeMessage(stdin); 
      const size = data.subarray(0, 4);
      const view = new DataView(size.buffer);
      const length = view.getUint32(0, true);
      sendMessage(encodeMessage(JSON.stringify({ length, size: [...size] })));
      const message = data.subarray(4);
// ...
```

```
#!/usr/bin/env -S /home/user/bin/qjs -m --std
// Read stdin to V8's d8, send to d8
// const stdin = os.system("./read_d8_stdin.js", [`/proc/${pid.replace(/\D+/g, "")}/fd/0`]);
function read_d8_stdin([, path] = scriptArgs) {
  try {
    // Uint32Array to read message length into
    const size = new Uint32Array(1);
    const err = {
      errno: 0
    };
    // Open /proc/PID/fd/0 for reading
    const pipe = std.open(
      path,
      "rb",
      err,
    );
    if (err.errno !== 0) {
      throw `${std.strerror(err.errno)}: ${path}`;
    }
    // Read /proc/PID/fd/0 into Uint32Array
    pipe.read(size.buffer, 0, 4);
    // Uint8Array to read message into
    const output = new Uint8Array(size[0]);
    // Read /proc/PID/fd/0 into Uint8Array
    pipe.read(output.buffer, 0, output.length);
    // Uint8Array containing Uint32Array buffer in Uint8Array and Uint8Array containing message
    const data = new Uint8Array([...new Uint8Array(size.buffer), ...output]);
    // Write Uint8Array to d8's stdin
    std.out.write(data.buffer, 0, data.length);
    std.out.flush();
    // Exit and start again initiated from d8
    std.exit(0);
  } catch (e) {
    // Handle error
    const err = {
      errno: 0
    };
    const file = std.open("qjsErr.txt", "w", err);
    if (err.errno !== 0) {
      file.puts(JSON.stringify(err));
      file.close();
      std.exit(1);
    }
    file.puts(JSON.stringify(e));
    file.close();
    std.out.puts(JSON.stringify(err));
    std.exit(1);
  }
}

read_d8_stdin();
```

**Comment by saghul** [comment]
> Right now I'm using QuickJS to read standard input stream to `d8` with `os.system()`, to get around `readline()` not being capable of processing `Uint32Array` from Chrome.

Can you please elaborate? It's not clear to me what you are trying to do here.

**Reply by guest271314** [reply]
I'm using `d8` as a Native Messaging host https://developer.chrome.com/docs/extensions/mv3/nativeMessaging/#native-messaging-host-protocol

> Chrome starts each native messaging host in a separate process and communicates with it using standard input (`stdin`) and standard output (`stdout`). The same format is used to send messages in both directions; each message is serialized using JSON, UTF-8 encoded and is preceded with 32-bit message length in native byte order. The maximum size of a single message from the native messaging host is 1 MB, mainly to protect Chrome from misbehaving native applications. The maximum size of the message sent to the native messaging host is 4 GB.

`d8`'s `readline()`, that expects a string, and executes that string, doesn't want to work to read the "32-bit message length in native byte order". SpiderMonkey's `js` `readline()` does not exit after reading the essentially `Uint32Array` https://github.com/guest271314/native-messaging-spidermonkey-shell/blob/main/nm_spidermonkey.js#L10-L36

```
function getMessage() {
  // Call readline() N times to catch `\r\n\r\n"` from 2d port.postMessage()
  let stdin;
  while (true) {
    stdin = readline();
    if (stdin !== null) {
      break;
    }
  }
  
  let data = `${stdin}`.replace(/[\r\n]+|\\x([0-9A-Fa-f]{2,4})/gu, "")
    .replace(/[^A-Za-z0-9\s\[,\]\{\}:_"]+/igu, "")
    .replace(/^"rnrn/gu, "")
    .replace(/^[#\r\n\}_]+(?=\[)/gu, "")
    .replace(/^"(?=["\{]+)|^"(?!"$)/gu, "") 
    .replace(/^\[(?=\[(?!.*\]{2}$))/gu, "")
    .replace(/^\{(?!\}|.+\}$)/gu, "")
    .replace(/^[0-9A-Z]+(?=[\[\{"])/igu, "") 
    .replace(/^[\]\}](?=\[)/i, "")
    .trimStart().trim();
  // https://stackoverflow.com/a/52434176
  // let previous = redirect("length.txt");
  // putstr(data.length);
  // redirect(previous); // restore the redirection to stdout
  // os.file.writeTypedArrayToFile("input.txt", encodeMessage(data));
  return encodeMessage(data);
}
```

In QuickJS world I know we can `import` a C shared object file.  I'm not sure if that capability exists in `d8`, and I already have QuickJS working to read stardard input to `d8` and send `d8`s `stdin` back to `d8`. 

The question is mainly about whether to jus keep using QuickJS JavaScript engine to read standard input to V8 (`d8` shell), or try to extend V8's `d8` shell to read standard input into an `ArrayBuffer` with C++. I have some working C++ code that I can try to incorporate into `d8.cc`. It looks like I'll use `v8::ArrayBuffer::Allocator` https://chromium.googlesource.com/v8/v8/+/refs/tags/12.8.241/src/d8/d8.cc#138. However, I have not extended or compiled V8 yet, and QuickJS works just fine right now to get `stdin` of and send that `stdin` back to the original application. So, use JavaScript to extend another JavaScript engine, or dive in to extending V8 with C++? It's kind of a philosophical programming question I suppose.

```
// C++ Native Messaging host
// https://browserext.github.io/native-messaging/
// https://developer.chrome.com/docs/apps/nativeMessaging/
// https://www.reddit.com/user/Eternal_Weeb/
// guest271314, 2022
#include <iostream>
#include <vector>
using namespace std;

vector<uint8_t> getMessage() {
  uint32_t length = 0;
  size_t size = fread(&length, sizeof(length), 1, stdin);
  vector<uint8_t> message(length); 
  size = fread(message.data(), sizeof(*message.data()), message.size(), stdin);
  return message;
}

void sendMessage(const vector<uint8_t> &message) {
  const uint32_t length = message.size();
  fwrite(&length, sizeof(length), 1, stdout);
  fwrite(message.data(), message.size(), sizeof(*message.data()), stdout);
  fflush(stdout);
}

int main() {
  while (true) {
    sendMessage(getMessage());
  }
  return 0;
}
```

**Reply by guest271314** [reply]
I'll add the V8 folks  do not appear interested in extending `d8` shell to process stardard input to an `ArrayBuffer` https://groups.google.com/g/v8-users/c/NsnStT6bx3Y/m/Yr_Z1FwgAQAJ.

**Reply by saghul** [reply]
Rather than sending the binary data, did you consider sending a serialized version of the Uint32Array (JSON perhaps?) if you send that + `\n` then I suppose `readline` will work, and you can deserialize the payload on the other end.

**Reply by guest271314** [reply]
Tried that. `d8`s `readline()` is exiting right after reading the raw "32-bit message length in native byte order". `d8`'s `readline()` and SpiderMonkey's `readline()` don't work the same.

**Reply by saghul** [reply, marked as answer]
I am not familiar with either of those, but I guess they might interpret what they are reading as newline and stopping there. In that case sticking to QuickJS might be the right option, other than switching to another engine like Deno or Bun.

**Reply by guest271314** [reply]
@saghul FWIW https://github.com/guest271314/v8/commit/75ec4afc989e9452903ff271777c2f6b41501e89#diff-3b74c96d943457c2c7022c0d1c1fb076f9faaf40704a76450b563b4d4b0513ac
